cluster-autoscaler/simulator/dynamicresources/snapshot/patchset.go

```diff
@@ -0,0 +1,237 @@
+func newPatchSet[K comparable, V any](patches ...*patch[K, V]) *patchSet[K, V] {
+	return &patchSet[K, V]{
+		patches: patches,
+		cache:   make(map[K]*V),
+	}
+}
+
+// Fork adds a new, empty patch layer to the top of the stack.
+// Subsequent modifications will be recorded in this new layer.
+func (ps *patchSet[K, V]) Fork() {
+	ps.patches = append(ps.patches, newPatch[K, V]())
+}
+
+// Commit merges the topmost patch layer into the one below it.
+// If there's only one layer (or none), it's a no-op.
+func (ps *patchSet[K, V]) Commit() {
+	if len(ps.patches) <= 1 {
+		return
+	}
+
+	currentPatch := ps.patches[len(ps.patches)-1]
+	previousPatch := ps.patches[len(ps.patches)-2]
+	mergedPatch := mergePatchesInPlace(previousPatch, currentPatch)
+	ps.patches = ps.patches[:len(ps.patches)-1]
+	ps.patches[len(ps.patches)-1] = mergedPatch
+}
+
+// Revert removes the topmost patch layer.
+// Any modifications or deletions recorded in that layer are discarded.
+func (ps *patchSet[K, V]) Revert() {
+	if len(ps.patches) <= 1 {
+		return
+	}
+
+	lastPatch := ps.patches[len(ps.patches)-1]
+	ps.patches = ps.patches[:len(ps.patches)-1]
+
+	for key := range lastPatch.Modified {
+		delete(ps.cache, key)
+	}
+
+	for key := range lastPatch.Deleted {
+		delete(ps.cache, key)
+	}
+}
+
+// FindValue searches for the effective value of a key by looking through the patches
+// from top to bottom. It returns the value and true if found, or the zero value and false
+// if the key is deleted or not found in any patch.
+func (ps *patchSet[K, V]) FindValue(key K) (value V, exists bool) {
+	var zero V
+
+	if cachedValue, cacheHit := ps.cache[key]; cacheHit {
+		if cachedValue == nil {
+			return zero, false
+		}
+
+		return *cachedValue, true
+	}
+
+	value = zero
+	exists = false
+	for i := len(ps.patches) - 1; i >= 0; i-- {
+		patch := ps.patches[i]
+		if patch.IsDeleted(key) {
+			value = zero
+			exists = false
+			break
+		}
+		foundValue, ok := patch.Get(key)
+		if ok {
+			value = foundValue
+			exists = true
+			break
+		}
+	}
+
+	if exists {
+		ps.cache[key] = &value
+	} else {
+		ps.cache[key] = nil
+	}
+
+	return value, exists
+}
+
+// AsMap merges all patches into a single map representing the current effective state.
+// It iterates through all patches from bottom to top, applying modifications and deletions.
+// The cache is populated with the results during this process.
+func (ps *patchSet[K, V]) AsMap() map[K]V {
+	keysCount := ps.approximateKeysCount()
+	merged := make(map[K]V, keysCount)
+
+	for _, patch := range ps.patches {
+		for key, value := range patch.Modified {
+			merged[key] = value
+		}
+
+		for key := range patch.Deleted {
+			delete(merged, key)
+		}
+	}
+
+	for key, value := range merged {
+		ps.cache[key] = &value
+	}
+
+	return merged
+}
+
+// SetCurrent adds or updates a key-value pair in the topmost patch layer.
+func (ps *patchSet[K, V]) SetCurrent(key K, value V) {
+	lastPatch := ps.patches[len(ps.patches)-1]
```

Comment:
same nit as above, let's name all of these `currentPatch` instead of `lastPatch`